Source/Core/VideoCommon/TextureInfo.cpp

```diff
@@ -0,0 +1,325 @@
+
+  m_name_cached = true;
+}
+
+std::optional<std::string> TextureInfo::GetFullName() const
+{
+  if (!m_name_cached)
+    return std::nullopt;
+
+  return m_base_name + m_tlut_name + m_format_name;
+}
+
+std::string_view TextureInfo::GetBaseName() const
+{
+  return m_base_name;
+}
+
+std::string_view TextureInfo::GetTlutName() const
+{
+  return m_tlut_name;
+}
+
+std::string_view TextureInfo::GetFormatName() const
+{
+  return m_format_name;
+}
+
+const u8* TextureInfo::GetData() const
+{
+  return m_data;
+}
+
+const u8* TextureInfo::GetTlutAddress() const
+{
+  return m_tlut_address;
+}
+
+u32 TextureInfo::GetRawAddress() const
+{
+  return m_raw_address;
+}
+
+bool TextureInfo::IsFromTmem() const
+{
+  return m_from_tmem;
+}
+
+const u8* TextureInfo::GetTmemOddAddress() const
+{
+  return m_tmem_odd;
+}
+
+TextureFormat TextureInfo::GetTextureFormat() const
+{
+  return m_texture_format;
+}
+
+TLUTFormat TextureInfo::GetTlutFormat() const
+{
+  return m_tlut_format;
+}
+
+std::optional<u32> TextureInfo::GetPaletteSize() const
+{
+  return m_palette_size;
+}
+
+u32 TextureInfo::GetTextureSize() const
+{
+  return m_texture_size;
+}
+
+u32 TextureInfo::GetBlockWidth() const
+{
+  return m_block_width;
+}
+
+u32 TextureInfo::GetBlockHeight() const
+{
+  return m_block_height;
+}
+
+u32 TextureInfo::GetExpandedWidth() const
+{
+  return m_expanded_width;
+}
+
+u32 TextureInfo::GetExpandedHeight() const
+{
+  return m_expanded_height;
+}
+
+u32 TextureInfo::GetRawWidth() const
+{
+  return m_raw_width;
+}
+
+u32 TextureInfo::GetRawHeight() const
+{
+  return m_raw_height;
+}
+
+bool TextureInfo::HasMipMaps() const
+{
+  return !m_mip_levels.empty();
+}
+
+u32 TextureInfo::GetLevelCount() const
+{
+  return static_cast<u32>(m_mip_levels.size()) + 1;
+}
+
+const TextureInfo::MipLevel* TextureInfo::GetMipMapLevel(u32 level) const
+{
+  if (m_mip_levels.size() > level)
+    return &m_mip_levels.at(level);
+
+  return nullptr;
+}
+
+TextureInfo::MipLevel::MipLevel(u32 level, const TextureInfo& parent, bool from_tmem,
+                                const u8*& src_data, const u8*& ptr_even, const u8*& ptr_odd)
+{
+  m_raw_width = std::max(parent.GetRawWidth() >> level, 1u);
+  m_raw_height = std::max(parent.GetRawHeight() >> level, 1u);
+  m_expanded_width = Common::AlignUp(m_raw_width, parent.GetBlockWidth());
+  m_expanded_height = Common::AlignUp(m_raw_height, parent.GetBlockHeight());
+
+  // load mips - TODO: Loading mipmaps from tmem is untested!
+
+  m_texture_size = TexDecoder_GetTextureSizeInBytes(m_expanded_width, m_expanded_height,
+                                                    parent.GetTextureFormat());
+
+  const u8*& data = from_tmem ? ((level % 2) ? ptr_odd : ptr_even) : src_data;
+  m_data = data;
+  data += m_texture_size;
+}
+
+u32 TextureInfo::GetFullLevelSize() const
+{
+  u32 all_mips_size = 0;
+  for (std::size_t i = 0; i < m_mip_levels.size(); i++)
```

Comment:
Can't this be condensed into a ranged for loop like

```cpp
for (const auto& mip_map : m_mip_levels)
```

This also avoids the copy-by-value that currently occurs within the loop.